proposals/NNNN-shader-execution-reordering.md

```diff
@@ -0,0 +1,1388 @@
+strategies are likely able to reuse existing mechanisms to implement support
+for SER. No DXR runtime changes are necessary, since the proposed extension to
+the programming model is limited to HLSL and DXIL.
+
+## Detailed Design
+
+This section describes the HLSL additions for `HitObject` and `ReorderThread`
+in detail.
+The canonical use of these features involve changing a `TraceRay` call to the
+following sequence:
+
+```C++
+HitObject Hit = HitObject::TraceRay( ..., Payload );
+ReorderThread( Hit );
+HitObject::Invoke( Hit, Payload );
+```
+
+This snippet traces a ray and stores the result of traversal, intersection
+testing and anyhit shading in `Hit`. The call to `ReorderThread` improves
+coherence based on the information inside the `Hit`. Closesthit or miss
+shading is then invoked in a more coherent context.
+
+Note that this is a very basic example. Among other things, it is possible to
+query information about the hit to influence `ReorderThread` with additional
+hints. See [Separation of ReorderThread and HitObject::Invoke](#separation-of-reorderthread-and-hitobjectinvoke) 
+for more elaborate examples.
+
+### HitObject HLSL Additions
+
+```C++
+HitObject
+```
+
+The `HitObject` type encapsulates information about a hit or a miss. A
+`HitObject` is constructed using `HitObject::TraceRay`,
+`HitObject::FromRayQuery`, `HitObject::MakeMiss`, or `HitObject::MakeNop`. It
+can be used to invoke closesthit or miss shading using `HitObject::Invoke`,
+and to reorder threads for shading coherence with `ReorderThread`.
+
+The `HitObject` has value semantics, so modifying one `HitObject` will not
+impact any other `HitObject` in the shader. A shader can have any number of
+active `HitObject`s at a time. Each `HitObject` will take up some hardware
+resources to hold information related to the hit or miss, including ray,
+intersection attributes and information about the shader to invoke. The size
+of a `HitObject` is unspecified. As a consequence, `HitObject` cannot be
+stored in structured buffers. Ray payload structs must not have members of
+`HitObject` type. `HitObject` supports assignment (by-value copy) and can be
+passed as arguments to and returned from local functions. It is
+default-initialized to encode a NOP hit object (see `HitObject::MakeNop`).
+
+### Valid Shader Stages
+
+The `HitObject` type and related functions are available only in the following
+shader stages: `raygeneration`, `closesthit`, `miss` (the shader stages that
+may call `TraceRay`).
+
+---
+
+#### HitObject::TraceRay
+
+Executes traversal (including anyhit and intersection shaders) and returns the
+resulting hit or miss information as a `HitObject`. Unlike `TraceRay`, this
+does not execute closesthit or miss shaders. The resulting payload is not part
+of the `HitObject`. See 
+[Interaction with Payload Access Qualifiers](#interaction-with-payload-access-qualifiers)
+for details.
+
+```C++
+template<payload_t> 
+static HitObject HitObject::TraceRay( 
```

Comment:
We are interested in adding explicit communication from application in terms of what it wants (or needs) from the `TraceRay` call regarding reordering. I would assume we have to pass extra values to the call through extra args or through call attributes/decorations (similar to `[unroll]`) instead of relying on implicit behavior (which differs between vendors - some do implicitly sort/reorder, some don't).

This could be implemented using three possible values (using attribute form of expressing it in examples):

1. `PreferNoReorder` - the default value with the behavior according to the current spirit. Application doesn't need the reordering happen but the call still is a reorder point and implementation can decide to reorder. Note: This could be also simply spelled out in the spec to be the default if `Reorder` calls are present.
2. `Reorder` - with this value, it is equivalent to:
   ```c++
   HitObject hit = [PreferNoReorder]HitObject::TraceRay( ... );
   ReorderThread( hit ); 
   ```
   the value should be limited to shader stages supporting `ReorderThread(...)`.
3. `NoReorder` - the value means that implementation is required to _not_ reorder. This may be necessary for algorithm correctness in some cases (like, a wave remaining the same before and after the `TraceRay` call). Note: There should be a note mentioning that this _can_ degrade performance and thus should be only used for correctness, otherwise see `PreferNoReorder`.

Reply:
The proposed [Reorder] is equivalent to TraceRay followed by ReorderThread. We believe this could introduce confusion to an already complex API, especially regarding when and where reordering occurs. Instead, the implementation can reasonably perform straightforward analysis to recognize sequences like TraceRay/ReorderThread/Invoke, reducing the need for additional API complexity.

Regarding [NoReorder] we agree with @Flakebi about the feasibility of such an option. While RayQuery can be implemented like this (as you mention in a separate comment) it is not certain that any implementation can support calling anyhit and intersection shaders while not modifying active threads.

We can explicitly make HitObject::TraceRay and Invoke not reorder. Then the app can control reordering by adding or omitting ReorderThreads. If the app wants the implementation to decide, legacy TraceRay can be used. Note that this does not actually guarantee anything about active threads before/after the call. It is more about the trade-off between overhead and coherence.

Reply:
So you're saying that some implementations can support ray query, but can't execute the shaders per lane without a resort? That's an interesting limitation which is currently not expressed in the specification, as there seem to be implied uses where the thread assignment gets _not_ reordered by `TraceRay` even in the presence of any hit shaders being called by `TraceRay`. For example, if I want to keep quads together, you're saying there's no way to do that right now with `TraceRay` as explained in the SER specification?

My assumption is that any implementation support ray-query can maintain the thread assignment.

The complication in the "straightforward analysis" is that tooling may move instructions around between `TraceRay`, `ReorderThread`, `Invoke`, and some implementations may prefer to use the `ReorderThread` into either `TraceRay` _or_ `Invoke`, and then it's undecidable which one to do if there's a `ReorderThread` in the middle of it, even though the implementation may know that re-ordering in `TraceRay` for example will provide no benefit - and it can't communicate this. `ReorderThread` is a fine API if all re-ordering would be limited to just that API call, but the reality is that there are multiple re-order points, and the specification language for SER specifically uses comments to explain where no reordering should happen - which is unfortunately not something the compiler knows about. I.e. in samples like:

```
// Invoke unified surface shading. We are already hit-coherent so not worth
// reordering again.
HitObject::Invoke( hit, payload );
```

Reply:
> So you're saying that some implementations can support ray query, but can't execute the shaders per lane without a resort? That's an interesting limitation which is currently not expressed in the specification, as there seem to be implied uses where the thread assignment gets not reordered by TraceRay even in the presence of any hit shaders being called by TraceRay. For example, if I want to keep quads together, you're saying there's no way to do that right now with TraceRay as explained in the SER specification?

TraceRay is a reorder point so there are no guarantees about active threads afterwards. This doesn't necessarily imply that a full reorder will happen, but some amount of changes in active threads may still happen. As mentioned, RayQuery does not invoke shaders, so it is not exactly comparable to TraceRay that itself calls anyhit and intersection shaders.

> The complication in the "straightforward analysis" is that tooling may move instructions around between TraceRay, ReorderThread, Invoke

Could you clarify what the tooling is here? If you mean dxc I believe we can restrict movement of things like resource accesses.

> some implementations may prefer to [f]use the ReorderThread into either TraceRay or Invoke, and then it's undecidable which one to do if there's a ReorderThread in the middle of it, even though the implementation may know that re-ordering in TraceRay for example will provide no benefit - and it can't communicate this

Could you expand on this? If the driver knows that fusing with TraceRay will provide no benefit it seems logical that it should fuse with Invoke.

> ReorderThread is a fine API if all re-ordering would be limited to just that API call, but the reality is that there are multiple re-order points, and the specification language for SER specifically uses comments to explain where no reordering should happen - which is unfortunately not something the compiler knows about. I.e. in samples like:

Since there is no ReorderThread call before the Invoke, it makes sense to assume that no extra effort should be made to reorder the Invoke, beyond what is required for the underlying architecture. If the underlying architecture requires an extra reorder for the Invoke anyway, it may want to utilize information provided in the most recent ReorderThread call.

Reply:
To your starting point: That's exactly the problem. `TraceRay` _can_ be a reorder point, but if you happen to run on an implementation which doesn't, you can write code which relies on that and you can't express it any more. I don't like "implicit" defaults because implicit is platform dependent and thus hurts portability. I'm sure there's already code out there which assumes no reordering happens because there's only one any hit shader present for example and empirically it didn't. As to stories about implicit assumptions, `if (AMD) waveSize=64` was another implicit thing for quite some time :)

The problem is the tooling can move code across and there's no guarantee that running it through SPIR-V opt or not will not move around, and the problem I'm mentioning is not that the _driver_ knows that fusing with `TraceRay` has no benefit; it's only the app who knows this. I.e. in a case like:

```
TraceRay();
SomeCalculationWhichWillBeIncoherentNoMatterWhatReorderingHappensInTraceRay();
ReorderThread();
SomethingExpensiveWhichIsNowCoherent();
Invoke();
SomethingElseWhichRemainsCoherent();
```

The application may know reordering in `TraceRay` provides no possible benefit; but it can't tell the implementation. There are implementations which would _always_ reorder in `TraceRay` and `Invoke`, and _additionally_ in `ReorderThread` if requested by the application. Now you have an unsolvable problem: You have 3 re-order points, one of which is _explicit_ and two are _implicit_, and there's no way the application can tell you which of the implicit ones is "less valuable".

To the last point, "assuming no extra effort" is not something I like to assume, because assumptions _will_ vary between implementations and in the end we'll be just figuring out what the most common assumption was.

(Sorry, a bit out of order here :) )

As to moving around stuff: I'm not aware of DXC restricting moving of instructions around, nor, in the future, SPIR-V tools or custom processing paths. Knowing if `TraceRay` is assumed to be "useless" provides information to the compiler in terms of code motion opportunities - it can move things around freely as it can assume a sample call for example is the same cost before/after `TraceRay`; when it's a reordering `TraceRay`, moving a sample before it will be likely prohibitive if the state needs to get saved/restored.

My main concern is here that we'll end up with multiple implementations with multiple defaults as to when to sort and in which call, and what turns out to be best for an application will be determined by the platform they happen to develop on, and when they move from one implementation which happens to sort a lot in `TraceRay` by default they'll realize "oh `ReorderThread` doesn't seem to help here" and leave it out, then on the next implementation they'll be "Oh I really need `ReorderThread` to make this work", but that was because the second implementation didn't do any measurable reorder, etc. In the end we'll be all app-opting where/when to sort even though the application was written with certain assumptions in mind they simply could not communicate to us other than comments.

Reply:
Our proposed solution is to specify that HitObject::TraceRay and HitObject::Invoke will not perform heavyweight reordering. However, they still serve as reorder points, giving the driver flexibility in how execution is scheduled. Shaders that rely on having the same sets of active threads before and after these calls are not valid.

In your example, this means that the first incoherent calculation is assumed to not be worth reordering. The only reordering that may occur is related to the efficiency of the HitObject::TraceRay operation itself, which is left to the driver’s discretion. Shader code that benefits from and receive coherence is, by definition, the code that follows ReorderThread. It is then up to the driver to decide whether it is beneficial to fuse ReorderThread with Invoke, depending on the intervening code. My assumption is that it’s better not to fuse them if there is non-uniform resource access in between, as this makes it easier for the application to predict behavior based on code changes.

Regarding tooling, it is important to ensure that non-uniform resource access remains aligned with ReorderThread. Moving accesses across it will defeat its purpose.

Reply:
That requires us to specify it though: "TraceRay and Invoke will not perform 'heavy-weight' reordering", I assume only in the presence of `ReorderThreads`, and thus behave as-if we always specified the `[NoReorder]` flag on them (well, nearly, because you're saying not heavy-weight, which would have to be defined as well - what does heavyweight mean?) I think it's clear by this point that we're back to the original reason we wanted this expressed clearly and explicitly, to not rely on interpretation of what heavyweight means or not.

Which brings me back to the question I raised - are there implementations which simply _cannot_ guarantee that `TraceRay` maintains a 1:1 mapping? If so, that makes `[NoReorder]` impossible to implement, but I don't see why that should be the case given you can always fall back to inlining all shaders and using `RayQuery`, and most likely applications would only use that in cases where there is no reason to reorder to begin with (i.e. only one any hit shader.)

Reply:
> That requires us to specify it though

Yes, we can specify this clearly and it can be phrased in a better way. The important part is what the app conveys to the driver and what the driver is allowed to do in return. The app does not explicitly expect any reordering for HitObject::TraceRay and HitObject::Invoke but the driver is still allowed flexibility for scheduling the work efficiently. The driver can account for things like the number of shaders and how to run them, and whether to fuse ReorderThread with another call (which effectively moves the reorder point).

> Which brings me back to the question I raised - are there implementations which simply cannot guarantee that TraceRay maintains a 1:1 mapping?

Forcing 1:1 mapping is very restrictive and not something we'd be comfortable with. It is certainly not always the case today. It is also unclear why this would be useful given that the app has little control over thread arrangement to begin with.

Inlining is not generally a valid solution as shaders are compiled separately and pipeline creation should not consume excessive time for recompilation. It can also become problematic if there are many call sites that need inlining of the same shader(s).

Reply:
Regarding a strict `[NoReorder]`, agreed that we shouldn't allow the app to force a 1:1 thread mapping.

The driver can't always inline everything (consider a pipeline with many shaders and many `TraceRay` calls in nested CHS/Miss, and a large recursion limit, or early compilation of collections), and without inlining such a guarantee would put an unreasonable burden on the implementation.
Note that this issue is not just about the set of active threads immediately before/after a TraceRay, but also about reconvergence with other threads that did not execute the `TraceRay`, see the earlier example by @Flakebi. 

However, a `[PreferNoReorder]` hint on reorder points that tells the driver it shouldn't spend any additonal effort on reordering for coherence would be fine in that regard.

Reply:
The `Reorder points` section has been updated with a more detailed account for expected reordering behavior. Let me know if this aligns with our discussion or if there is still a need for attributes.

Reply:
In last week's meeting we talked about how the default behavior is no reorder, but implementations can allow reordering if explicitly specified by user. Hitobjecttraceray followed by hitobjectinvoke is a prefer no reorder, because no reorder thread is called directly. No attributes or parameters needed. 

Are we ok to close this comment?